cohttp-eio/src/reader.ml

```diff
@@ -0,0 +1,322 @@
+    incr_pos ~n:count rdr;
+    s)
+  else ""
+
+let take_bigstring : int -> Bigstringaf.t parser =
+ fun n rdr ->
+  try
+    ensure rdr n;
+    let s = copy rdr ~off:(pos rdr) ~len:n in
+    incr_pos ~n rdr;
+    s
+  with End_of_file -> fail "[take_bigstring] not enough input" rdr
+
+let take_bytes : int -> bytes parser =
+ fun n rdr ->
+  try
+    ensure rdr n;
+    let s = sub rdr ~off:(pos rdr) ~len:n in
+    incr_pos ~n rdr;
+    s
+  with End_of_file -> fail "[take] not enough input" rdr
+
+let take : int -> string parser =
+ fun n rdr ->
+  try
+    ensure rdr n;
+    let s = substring rdr ~off:(pos rdr) ~len:n in
+    incr_pos ~n rdr;
+    s
+  with End_of_file -> fail "[take] not enough input" rdr
+
+let take_till f = take_while (fun c -> not (f c))
+
+let rec many : 'a parser -> 'a list parser =
+ fun p rdr ->
+  try
+    let a = p rdr in
+    a :: many p rdr
+  with Parse_failure _ | End_of_file -> []
+
+let rec many_till : 'a parser -> _ parser -> 'a list parser =
+ fun p t rdr ->
+  try
+    let _ = t rdr in
+    let a = p rdr in
+    a :: many_till p t rdr
+  with Parse_failure _ -> []
+
+let skip f rdr =
+  ensure rdr 1;
+  let c = unsafe_get rdr (pos rdr) in
+  if f c then incr_pos rdr else fail "[skip]" rdr
+
+let skip_while f rdr =
+  let count = count_while rdr f in
+  incr_pos ~n:count rdr
+
+let rec skip_many p rdr =
+  match p rdr with _ -> skip_many p rdr | exception Parse_failure _ -> ()
+
+(* Builtin readers *)
+
+let token =
+  take_while1 (function
+    | '0' .. '9'
+    | 'a' .. 'z'
+    | 'A' .. 'Z'
+    | '!' | '#' | '$' | '%' | '&' | '\'' | '*' | '+' | '-' | '.' | '^' | '_'
+    | '`' | '|' | '~' ->
+        true
+    | _ -> false)
+
+let ows = skip_while (function ' ' | '\t' -> true | _ -> false)
+let crlf = string "\r\n"
+let is_cr = function '\r' -> true | _ -> false
+let space = char '\x20'
+let p_meth = token <* space >>| Http.Method.of_string
+let p_resource = take_while1 (fun c -> c != ' ') <* space
+
+let p_version =
+  string "HTTP/1." *> any_char <* crlf >>= function
+  | '1' -> return `HTTP_1_1
+  | '0' -> return `HTTP_1_0
+  | v -> fail (Format.sprintf "Invalid HTTP version: %C" v)
+
+let header =
+  lift2
+    (fun key value -> (key, value))
+    (token <* char ':' <* ows)
+    (take_till is_cr <* crlf)
+
+let http_headers =
+  let cons x xs = x :: xs in
+  fix (fun headers ->
+      let _emp = return [] in
+      let _rec = lift2 cons header headers in
+      peek_char >>= function '\r' -> _emp | _ -> _rec)
+  >>| Http.Header.of_list
+  <* crlf
+
+let[@warning "-3"] http_request t =
+  match end_of_input t with
+  | true -> Stdlib.raise_notrace End_of_file
+  | false ->
+      let meth = p_meth t in
+      let resource = p_resource t in
+      let version = p_version t in
+      let headers = http_headers t in
+      let encoding = Http.Header.get_transfer_encoding headers in
+      commit t;
+      { Http.Request.meth; resource; version; headers; scheme = None; encoding }
```

Comment:
Since we were talking about scheme, why here we set it to `None`? We don't have a way to populate it?
Is the warning due to the use of encoding?

Reply:
> Since we were talking about scheme, why here we set it to None? We don't have a way to populate it?

I am not sure if there is a reliable way to determine a value for the scheme. For example, a web app using cohttp-eio could be behind a proxy server where the ssl/tls terminate or it could be a normal socket connection without ssl/tls.

> Is the warning due to the use of encoding?

Yes, dune gives a build error during dev without the attribute.